### PR TITLE
Deal with useless params in prompt

### PR DESCRIPTION
**Pull request by Henry-Pulver**
Summation of changes incl new version of `chat-ui` leads to the below:

https://github.com/Superflows-AI/superflows/assets/33871096/46cedf4d-84f4-4991-b0e0-2f8eca13f7ed

### CODE REVIEW

File: lib/actionUtils.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        typeof b[key] === "object" &&
+        b[key] !== null &&
+        !Array.isArray(b[key]) &&
+        out[key] &&
+        typeof out[key] === "object" &&
+        !Array.isArray(out[key])
+      ) {
+        out[key] = combineParams(out[key], b[key]);
+      } else {
+        // @ts-ignore
```
Comment (neuromantic99) [inline lib/actionUtils.ts:106]: ts-ignore always freaks me out but i think the typing makes sense here and i couldn't find any other way to do it

File: __tests__/lib/actionUtils.test.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+describe("isChoiceRequired", () => {
+  it("nothing to fill 1", () => {
+    const out = isChoiceRequired(nothingToFillSchema1);
+    expect(out).toBe(true);
+  });
+  it("nothing to fill 2", () => {
+    const out = isChoiceRequired(nothingToFillSchema2);
+    expect(out).toBe(true);
+  });
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply __tests__/lib/actionUtils.test.ts:106]: I read further down and the variable names make sense in that context but not in the above.

DSS (do something smart)

File: lib/actionUtils.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  }
+
+  return filledObject;
+}
+
+export function fillNoChoiceRequiredParams<
+  Params extends Record<string, unknown>,
+  Output extends Params
+>(paramsFromAI: Params, schema: OpenAPIV3_1.SchemaObject): Output {
```
Comment (neuromantic99) [inline lib/actionUtils.ts:73]: I think this either needs a comment or a new name e.g. `combineAIandAutomaticParams`, that's rubbish but you get my drift.
 it took me a while to figure out that you pull out "fillOutable" params from the schema then combine them with the params from the AI

File: __tests__/lib/actionUtils.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Henry-Pulver) [reply __tests__/lib/actionUtils.test.ts:106]: DSS (do something smart) :joy:

File: lib/actionUtils.ts (hunk 1 above)
Comment (Henry-Pulver) [reply lib/actionUtils.ts:106]: Yep, I asked GPT and it couldn't explain why this is happening (or what we could do to solve it) other than basically setting the return type to something less restrictive (ie worse) like `Record<string, any>`

File: lib/actionUtils.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+function combineParams<
+  ParamsA extends Params,
+  ParamsB extends Params,
+  OutParams extends ParamsA & ParamsB
+>(a: ParamsA, b: ParamsB): OutParams {
+  /** This function combines two objects, with b overwriting a in
+   * the case of a clash. **/
+  // TODO: Doesn't deal with arrays yet, only deals with nested objects
```
[the remark this answers is not in the document]
Comment (Henry-Pulver) [reply lib/actionUtils.ts:95]: > is this not a major issue? Or is this only applied currently in special cases for IF where there are no arrays?
Basically yes

Added a ticket to make this work for arrays too: https://learney.atlassian.net/browse/SF-2093

**Pull request merged**